internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -78,27 +85,86 @@ func (p *processRunner) Close() error {
 }
 
 type commandWrapper struct {
-	cmd              *exec.Cmd
+	launcher         *cap.Launcher
+	ctty             optional.Optional[int]
+	stdin            uintptr
+	stdout           uintptr
+	stderr           uintptr
 	afterStart       func()
 	afterTermination func() error
 }
 
+func dropCaps(droppedCapabilities []string, launcher *cap.Launcher) error {
+	droppedCaps := strings.Join(droppedCapabilities, ",")
+
+	if droppedCaps != "" {
+		caps := strings.Split(droppedCaps, ",")
+		dropCaps := xslices.Map(caps, func(c string) cap.Value {
+			capability, capErr := cap.FromName(c)
+			if capErr != nil {
+				fmt.Printf("failed to parse capability: %s", capErr)
+			}
+
+			return capability
+		})
+
+		iab := cap.IABGetProc()
+		if err := iab.SetVector(cap.Bound, true, dropCaps...); err != nil {
+			return fmt.Errorf("failed to set capabilities: %w", err)
+		}
+
+		launcher.SetIAB(iab)
+	}
+
+	return nil
+}
+
+// This callback is run in the thread before executing child process.
+func beforeExecCallback(pa *syscall.ProcAttr, data interface{}) error {
+	wrapper, ok := data.(*commandWrapper)
+	if !ok {
+		return fmt.Errorf("failed to get command info")
+	}
+
+	ctty, cttySet := wrapper.ctty.Get()
+	if cttySet {
+		if pa.Sys == nil {
+			pa.Sys = &syscall.SysProcAttr{}
+		}
+
+		pa.Sys.Ctty = ctty
+		pa.Sys.Setsid = true
+		pa.Sys.Setctty = true
+	}
+
+	pa.Files = []uintptr{
+		wrapper.stdin,
+		wrapper.stdout,
+		wrapper.stderr,
+	}
+
+	// TODO: use pa.Sys.CgroupFD here when we can be sure clone3 is available
+	return nil
+}
+
 //nolint:gocyclo
 func (p *processRunner) build() (commandWrapper, error) {
-	args := []string{
-		fmt.Sprintf("-name=%s", p.args.ID),
-		fmt.Sprintf("-dropped-caps=%s", strings.Join(p.opts.DroppedCapabilities, ",")),
-		fmt.Sprintf("-cgroup-path=%s", cgroup.Path(p.opts.CgroupPath)),
-		fmt.Sprintf("-oom-score=%d", p.opts.OOMScoreAdj),
-		fmt.Sprintf("-uid=%d", p.opts.UID),
-	}
+	wrapper := commandWrapper{}
 
-	args = append(args, p.args.ProcessArgs...)
+	env := slices.Concat([]string{"PATH=" + constants.PATH}, p.opts.Env, os.Environ())
+	launcher := cap.NewLauncher(p.args.ProcessArgs[0], p.args.ProcessArgs, env)
 
-	cmd := exec.Command("/sbin/wrapperd", args...)
+	// forkExec fails if we do setuid
+	// if p.opts.UID > 0 {
+	// 	launcher.SetUID(int(p.opts.UID))
+	// }
```

Comment:
And if we move this to system process args, then Launcher's validation triggers. Going to take a look later.